man: add `sssd-passkey.5`

ikerexxe opened the pull request:
Include a new man page for passkey to explain the behaviour of `user_verification` option in the different scenarios. It is a complex option, so it has been decided to add a table to simplify its understanding.

ikerexxe commented:
@justin-stephenson and @sumit-bose here is the documentation that we agreed to create yesterday. Please take a look at it when you have some time.

justin-stephenson commented:
Thanks very much for working on this @ikerexxe

To understand more clearly I did some testing of our current handling of passkey user-verification, specifically with kerberos-enabled passkey auth with IPA.

When IPA `require-user-verification == true` and my yubikey device has **NO PIN** set: device auth will fail as we always try to send a PIN to the passkey_child. If 'enter' is pressed without typing any characters then sssd falls back from passkey to password auth, instead of sending an empty PIN string.

When IPA `require-user-verification == false` and my yubikey device has a **PIN set**: device auth will fail as no PIN is sent to SSSD and then to the passkey_child.

Is this the expected behavior when a device does not match the IPA server side policy to require user verification?

justin-stephenson commented on src/man/sssd-passkey.5.xml:
```diff
@@ -0,0 +1,105 @@
+                        <para>
+                            The configuration option from <filename>sssd.conf
+                            </filename> has higher priority than the LDAP
+                            server and will override them when conflicts occur.
```
I haven't tested local passkey auth but at least in the kerberos case this should not be true. See https://github.com/SSSD/sssd/blob/master/src/responder/pam/pamsrv_passkey.c#L175-L180

ikerexxe replied:
Didn't we agree in the design phase that `sssd.conf` takes precedence over IPA settings?

sumit-bose replied:
Hi,

I think at least in the Kerberos case it does not make sense to overwrite what we got form the KDC because iiuc the reply from the device which we send back to the KDC will contain information if user verification was used or not. If the KDC requires user verification and we would allow locally a device without a PIN the reply would be rejected by the KDC.

Maybe, it would be easier to say that with the IPA provider the server side settings will be used (since IPA is currently the only provider which supports server side settings) and for every other provider the settings from sssd.conf are used?

bye,
Sumit

ikerexxe replied:
> Maybe, it would be easier to say that with the IPA provider the server side settings will be used (since IPA is currently the only provider which supports server side settings) and for every other provider the settings from sssd.conf are used?

I think this is the best solution, as it enables us to have a consistent approach and we do not confuse users/sysadmins.

sumit-bose replied:
Ok, I'm fine with this.

justin-stephenson replied:
I'm also okay with this.

ikerexxe replied:
I have updated the text to express what was discussed in this issue

ikerexxe commented:
> When IPA `require-user-verification == true` and my yubikey device has **NO PIN** set: device auth will fail as we always try to send a PIN to the passkey_child.

Yes, that's the expected behaviour and that's what I intended to convey with:
| sssd.conf | Device | Result |
|---|---|---|
| True | User verification is not configured  | Authentication fails  |

I know I mention `sssd.conf` instead of IPA, but I wanted to simplify the table and taking into account that `sssd.conf` takes precedence over IPA I thought this was the best to achieve it. If this is not understood, let me know and we will try to improve the text.

> If 'enter' is pressed without typing any characters then sssd falls back from passkey to password auth, instead of sending an empty PIN string.

Should we mention this in the man page?

> When IPA `require-user-verification == false` and my yubikey device has a **PIN set**: device auth will fail as no PIN is sent to SSSD and then to the passkey_child.

This doesn't align with the text I wrote. I based my understanding in the local case, where https://github.com/SSSD/sssd/blob/master/src/passkey_child/passkey_child_credentials.c#L133 is executed, and thus, the PIN is requested. Taking into account that the device won't allow the authentication without the PIN I think this is something that we need to improve in the online/kerberos workflow. As we'll run the `preflight` option in the preauthentication phase, we can use it to prepare for this use case and request the PIN. WDYT?

sumit-bose commented on src/man/sssd-passkey.5.xml:
```diff
@@ -0,0 +1,105 @@
+                        <row>
+                            <entry>True</entry>
+                            <entry>User verification is not configured</entry>
+                            <entry>Authentication fails</entry>
```
Hi,

I think might be more consistent to handle this case as if no (suitable) device is available at the pre-flight check. Then users have the chance to insert another device which requires a PIN before they enter the PIN. 

bye,
Sumit

sumit-bose replied:
Hi,

no, the pre-flight is is internal and should not be mentioned in the man page explicitly. What I meant is that "Authentication fails" give the impression that authentication fails immediately but I guess in reality what happens in the authentication phase is more a "User verification is requested but it is expected that authentication fails if no other device is inserted in between".

bye,
Sumit

ikerexxe replied:
Can you take a look at the updated text? I think I have reflected what you mention, but I would prefer to have a confirmation.

justin-stephenson commented:
> > If 'enter' is pressed without typing any characters then sssd falls back from passkey to password auth, instead of sending an empty PIN string.
> 
> Should we mention this in the man page?

We can mention this.
> 
> > When IPA `require-user-verification == false` and my yubikey device has a **PIN set**: device auth will fail as no PIN is sent to SSSD and then to the passkey_child.
> 
> This doesn't align with the text I wrote. I based my understanding in the local case, where https://github.com/SSSD/sssd/blob/master/src/passkey_child/passkey_child_credentials.c#L133 is executed, and thus, the PIN is requested. Taking into account that the device won't allow the authentication without the PIN I think this is something that we need to improve in the online/kerberos workflow. As we'll run the `preflight` option in the preauthentication phase, we can use it to prepare for this use case and request the PIN. WDYT?

Okay, yes I agree that this is a better approach. If UV is false (whether sssd.conf or IPA require-user-verification) and the device has a PIN set then SSSD will automatically determine this when it queries if the device has a PIN set and prompt for a PIN. This is also something to clearly state in the man page IMO.

justin-stephenson commented:
> > This doesn't align with the text I wrote. I based my understanding in the local case, where https://github.com/SSSD/sssd/blob/master/src/passkey_child/passkey_child_credentials.c#L133 is executed, and thus, the PIN is requested. Taking into account that the device won't allow the authentication without the PIN I think this is something that we need to improve in the online/kerberos workflow. As we'll run the `preflight` option in the preauthentication phase, we can use it to prepare for this use case and request the PIN. WDYT?
> 
> Okay, yes I agree that this is a better approach. If UV is false (whether sssd.conf or IPA require-user-verification) and the device has a PIN set then SSSD will automatically determine this when it queries if the device has a PIN set and effectively re-enable UV for this auth(prompt for a PIN). This is also something to clearly state in the man page IMO.

I suppose this will need to be fixed in the code as part of https://github.com/SSSD/sssd/pull/7631 or my SSSD PR when it comes.

ikerexxe commented:
@justin-stephenson I updated the text to include the password fallback and the user verification from the device. Take a look at it when you have some time.

> I suppose this will need to be fixed in the code as part of #7631 or my SSSD PR when it comes.

Yes, that would be the best approach.

justin-stephenson commented:
> @justin-stephenson I updated the text to include the password fallback and the user verification from the device. Take a look at it when you have some time.

Looks good to me, thank you Iker.
> 
> > I suppose this will need to be fixed in the code as part of #7631 or my SSSD PR when it comes.
> 
> Yes, that would be the best approach.

alexey-tikhonov commented:
@ikerexxe, please set backport labels.

ikerexxe commented:
> @ikerexxe, please set backport labels.

I'm taking this PR as a forum to agree on the documentation. Once we all have the same understanding I'll close it and include the commit in https://github.com/SSSD/sssd/pull/7631, as we'll need to adapt the behaviour a little bit to match the documentation.

sumit-bose commented on src/man/sssd-passkey.5.xml:
```diff
@@ -0,0 +1,120 @@
+                    <listitem>
+                        <para>
+                            Enable or disable the user verification (i.e. PIN,
+                            fingerprint) during authentication.
```
Maybe this sentence can be made more specific by saying "Enable or disable the requirement for user verification (i.e. PIN, fingerprint) on the passkey device during authentication." ?

ikerexxe replied:
Done

sumit-bose commented on src/man/sssd-passkey.5.xml:
```diff
@@ -0,0 +1,120 @@
+                            higher priority than <filename>sssd.conf
+                            </filename>. For every other provider the settings
+                            from <filename>sssd.conf</filename> have higher
+                            priority.
```
I still think "priority" should be avoided here. If I understand it correctly in the IPA case only the server settings are used and for all other cases there are no server settings and only the settings from `sssd.conf` will be used. So there is no priority or ordering because there is always only a single source.

ikerexxe replied:
It's a little bit more complicated since we have the offline authentication workflow. What do you think about the next proposal?
```
For online authentication with the IPA provider, where a Kerberos ticket is granted, server-side settings are applied. For other cases, such as offline authentication with the IPA provider or any authentication from another provider, the settings in sssd.conf are used.
```

sumit-bose replied:
Hi,

I think the IPA provider should use the server-side settings even for the offline case if passkey authentication is available online. Otherwise it would be the same odd inconsistency we tried to avoid with `local_auth_policy = match` with respect to the available authentication methods.

Only if passkey authentication is not available online in the IPA case and is enable locally by setting `local_auth_policy = enable:passkey` the setting from `sssd.conf` can be used.

In a mixed case where users from a trusted AD forest have the required attributes for passkey authentication set and ``local_auth_policy = enable:passkey` must be set to allow passkey authentication for those users I think the server-side setting for IPA users should be applied as well.

bye,
Sumit

ikerexxe replied:
So what about
```
If the IPA provider is used, passkey authentication is not available online and
local_auth_policy = enable:passkey is set, then the settings from sssd.conf are used.
For any other IPA provider case, server-side settings are applied. For other cases, such
as any authentication from another provider, the settings in sssd.conf are used.
```
?

And I fear I must have lost some case :disappointed_relieved:

sumit-bose replied:
Hi,

I think it covers all with some aspects only implicit. But I would suggest to try to reorder the sentences so that the cases where sssd.conf is used are following each other. So maybe
```
If the IPA provider is used and online, Kerberos based passkey authentication
is available the server-side settings are applied for all passkey usages. For
other cases the settings in sssd.conf are used. This includes passkey
authentication with any other provider and with the IPA provider in case online,
Kerberos based passkey authentication is not available and local passkey
authentication is allowed by the option `local_auth_policy = enable:passkey`.
```

bye,
Sumit

ikerexxe replied:
I changed some minor parts of the text, but overall it's almost the same. Thanks!

sumit-bose commented on src/man/sssd-passkey.5.xml:
```diff
@@ -0,0 +1,120 @@
+                        </para>
+                        <para>
+                            In the case of the <filename>sssd.conf</filename>
+                            and the passkey device options the situation is
```
I would like to suggest to start the sentence with "The interaction of the `user_verification` option and the passkey device option is explained ...`" and label the first column `user_verification`.

ikerexxe replied:
Done

sumit-bose commented on src/man/sssd-passkey.5.xml:
```diff
@@ -0,0 +1,120 @@
+                            User verification is requested; however, the
+                            authentication is expected to fail if the device is
+                            not replaced with a device where authentication is
+                            not configured during the authentication process.
```
Shouldn't this be " ... replaced with a device where user verification _is_ configured ..."?

ikerexxe replied:
Yes, you are right

sumit-bose replied:
I still think it would be better to say "... where user verification is configured ..." instead of "... where authentication is configured ..." to make a more clear distinction between the authentication the user is doing against the system to log in and the means needed to prove that the user is the owner of the passkey device.

ikerexxe replied:
You are right, I think I mistyped there

sumit-bose commented on src/man/sssd-passkey.5.xml:
```diff
@@ -0,0 +1,120 @@
+                        </tbody></tgroup></informaltable>
+                        <para>
+                            If 'enter' is pressed without typing any
+                            characters, then sssd falls back from passkey to
```
I would add " ... at the PIN prompt for user verification ..." and see above about "SSSD".

ikerexxe replied:
> and see above about "SSSD".

I don't understand what you mean by this?

sumit-bose replied:
I meant writing "SSSD" in upper-case.

ikerexxe replied:
Done

ikerexxe commented:
Hi @sumit-bose,

Are you ok with the current proposal? If so, I'd like to close this PR and move the changes to https://github.com/SSSD/sssd/pull/7631 to have everything centralized in one place.

ikerexxe commented:
> thank you for the reminder, I just added a comment to one of the conversations, but feel free to ignore it, if you think the current sentence is clear enough. Otherwise I do not have further comments. Please go on and move the changes to #7631.

Updated, so I'm closing this PR and moving the changes to the other one. Thank you both for your reviews!
